internal/rate/sliding_log.go

```diff
@@ -0,0 +1,144 @@
+package rate
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/go-redis/redis/v8"
+
+	"github.com/TykTechnologies/tyk/storage"
+)
+
+type SlidingLog struct {
+	conn     redis.UniversalClient
+	pipeline bool
+}
+
+func NewSlidingLog(cluster storage.Handler, pipeline bool) (*SlidingLog, error) {
+	conn := new(redis.UniversalClient)
+	if err := cluster.As(conn); err != nil {
+		return nil, err
+	}
+
+	return NewSlidingLogRedis(*conn, pipeline), nil
+}
+
+func NewSlidingLogRedis(conn redis.UniversalClient, pipeline bool) *SlidingLog {
+	return &SlidingLog{
+		conn:     conn,
+		pipeline: pipeline,
+	}
+}
+
+// SetCount will trim the rolling window log, add an item and return the count of the items in a window before the add.
+func (r *SlidingLog) SetCount(ctx context.Context, keyName string, per int64) (int64, error) {
+	now := time.Now()
+	onePeriodAgo := now.Add(time.Duration(-1*per) * time.Second)
+
+	var res *redis.IntCmd
+
+	pipeFn := func(pipe redis.Pipeliner) error {
+		pipe.ZRemRangeByScore(ctx, keyName, "-inf", strconv.Itoa(int(onePeriodAgo.UnixNano())))
+		res = pipe.ZCard(ctx, keyName)
+
+		element := redis.Z{
+			Score: float64(now.UnixNano()),
+		}
+
+		element.Member = strconv.Itoa(int(now.UnixNano()))
+
+		pipe.ZAdd(ctx, keyName, &element)
+		pipe.Expire(ctx, keyName, time.Duration(per)*time.Second)
```

Comment:
do we need this expiry?

Reply:
Yes, but not on every request. It's preexisting and can be optimized further